server/schedulers/shuffle_region.go

```diff
@@ -125,31 +125,48 @@ func (s *shuffleRegionScheduler) Schedule(cluster opt.Cluster) []*operator.Opera
 
 func (s *shuffleRegionScheduler) scheduleRemovePeer(cluster opt.Cluster) (*core.RegionInfo, *metapb.Peer) {
 	stores := cluster.GetStores()
+	exclude := make(map[uint64]struct{})
+	excludeFilter := filter.NewExcludedFilter(s.GetType(), exclude, nil)
+
+	for {
+		source := s.selector.SelectSource(cluster, stores, excludeFilter)
+		if source == nil {
+			schedulerCounter.WithLabelValues(s.GetName(), "no-source-store").Inc()
+			return nil, nil
+		}
 
-	source := s.selector.SelectSource(cluster, stores)
-	if source == nil {
-		schedulerCounter.WithLabelValues(s.GetName(), "no-source-store").Inc()
-		return nil, nil
-	}
+		var region *core.RegionInfo
+		if s.conf.IsRoleAllow(roleFollower) {
+			region = cluster.RandFollowerRegion(source.GetID(), s.conf.GetRanges(), opt.HealthRegion(cluster), opt.ReplicatedRegion(cluster))
+		}
+		if region == nil && s.conf.IsRoleAllow(roleLeader) {
+			region = cluster.RandLeaderRegion(source.GetID(), s.conf.GetRanges(), opt.HealthRegion(cluster), opt.ReplicatedRegion(cluster))
+		}
+		if region == nil && s.conf.IsRoleAllow(roleLearner) {
+			region = cluster.RandLearnerRegion(source.GetID(), s.conf.GetRanges(), opt.HealthRegion(cluster), opt.ReplicatedRegion(cluster))
+		}
+		if region != nil {
+			return region, region.GetStorePeer(source.GetID())
+		}
 
-	region := cluster.RandFollowerRegion(source.GetID(), s.conf.Ranges, opt.HealthRegion(cluster))
-	if region == nil {
-		region = cluster.RandLeaderRegion(source.GetID(), s.conf.Ranges, opt.HealthRegion(cluster))
-	}
-	if region == nil {
+		exclude[source.GetID()] = struct{}{}
 		schedulerCounter.WithLabelValues(s.GetName(), "no-region").Inc()
-		return nil, nil
 	}
-
-	return region, region.GetStorePeer(source.GetID())
 }
 
-func (s *shuffleRegionScheduler) scheduleAddPeer(cluster opt.Cluster, filter filter.Filter) *metapb.Peer {
-	stores := cluster.GetStores()
+func (s *shuffleRegionScheduler) scheduleAddPeer(cluster opt.Cluster, region *core.RegionInfo, oldPeer *metapb.Peer) *metapb.Peer {
+	var scoreGuard filter.Filter
+	if cluster.IsPlacementRulesEnabled() {
+		scoreGuard = filter.NewRuleFitFilter(s.GetName(), cluster, region, oldPeer.GetStoreId())
+	} else {
+		scoreGuard = filter.NewDistinctScoreFilter(s.GetName(), cluster.GetLocationLabels(), cluster.GetRegionStores(region), cluster.GetStore(oldPeer.GetStoreId()))
```

Comment:
Is it different from the original behavior?

Reply:
yes, different. Priviously it does not consider any constraints.